src/audio/google_audio_post_processing.c

```diff
@@ -0,0 +1,497 @@
+// SPDX-License-Identifier: BSD-3-Clause
+//
+// Copyright(c) 2022 Google LLC.
+//
+// Author: Kehuang Li <kehuangli@google.com>
+
+#include <sof/audio/audio_stream.h>
+#include <sof/audio/buffer.h>
+#include <sof/audio/component.h>
+#include <sof/audio/format.h>
+#include <sof/audio/pipeline.h>
+#include <sof/compiler_attributes.h>
+#include <sof/lib/alloc.h>
+#include <stdint.h>
+#include <stdlib.h>
+
+#include <google_audio_post_processing.h>
+
+#define GOOGLE_AUDIO_POST_PROCESSING_SAMPLERATE 48000
+
+/* fd48f000-c316-4ec2-9ff8-edb4efa3f52c */
+DECLARE_SOF_RT_UUID("google-audio-post-processing", gapp_uuid, 0xfd48f000, 0xc316, 0x4ec2, 0x9f,
+		    0xf8, 0xed, 0xb4, 0xef, 0xa3, 0xf5, 0x2c);
+DECLARE_TR_CTX(gapp_tr, SOF_UUID(gapp_uuid), LOG_LEVEL_INFO);
+
+/* GAPP configuration & model calibration data for tuning/debug */
+#define SOF_GAPP_CONFIG 0
+#define SOF_GAPP_MODEL 1
+#define SOF_GAPP_DATA 2
+
+struct comp_data {
+	GoogleAudioPostProcessingState *state;
+	struct comp_data_blob_handler *tuning_handler;
+	uint32_t config[2];
+	int channel_volume[SOF_IPC_MAX_CHANNELS];
+	int num_channels;
+	bool has_new_volume;
+	struct GoogleAudioPostProcessingBuffer buf_in;
+	struct GoogleAudioPostProcessingBuffer buf_out;
+};
+
+static int gapp_setup(struct comp_dev *dev, struct comp_data *cd)
+{
+	uint8_t *config;
+	size_t config_size;
+
+	config = comp_get_data_blob(cd->tuning_handler, &config_size, NULL);
+	return GoogleAudioPostProcessingSetup(cd->state, cd->num_channels, dev->frames,
+					      cd->channel_volume[0], config, config_size);
+}
+
+static int gapp_set_volume(struct comp_data *cd)
+{
+	int ret = 0;
+
+	if (cd->has_new_volume) {
+		ret = GoogleAudioPostProcessingSetVol(cd->state, cd->channel_volume,
+						      cd->num_channels);
+		if (ret >= 0)
+			cd->has_new_volume = false;
+	}
+
+	return ret;
+}
+
+static int gapp_ctrl_set_val(struct comp_dev *dev, struct sof_ipc_ctrl_data *cdata)
+{
+	struct comp_data *cd = comp_get_drvdata(dev);
+	int j;
+	int ch;
+	uint32_t val;
+
+	switch (cdata->cmd) {
+	case SOF_CTRL_CMD_VOLUME:
+		comp_info(dev, "gapp_ctrl_set_val(): SOF_CTRL_CMD_VOLUME, comp_id = %u",
+			  cdata->comp_id);
```

Comment:
wouldn't `comp_dbg()` be enough? Particularly since you only log the fact of this command being sent, not even any volume values?

Reply:
Done.